### PR TITLE
[MI-3072]:Converted the LHS APIs into GraphQL

### PR DESCRIPTION
**Pull request by Kshitij-Katiyar**

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -91,3 +92,5 @@ require (
 	gopkg.in/yaml.v2 v2.4.0 // indirect
 	gopkg.in/yaml.v3 v3.0.1 // indirect
 )
+
+require github.com/shurcooL/graphql v0.0.0-20220606043923-3cf50f8a0a29 // indirect
```
Comment (ayusht2810) [inline go.mod]: should this be indirect dependency?

File: server/plugin/api.go [hunk 2]
```diff
@@ -970,27 +938,24 @@ func (p *Plugin) createIssueComment(c *UserContext, w http.ResponseWriter, r *ht
...
-	config := p.getConfiguration()
-
-	githubClient := p.githubConnectUser(c.Context.Ctx, c.GHInfo)
+func (p *Plugin) getLHSData(c *UserContext) ([]*github.Issue, []*github.Issue, []*github.Issue) {
+	graphQLClient := p.graphQLConnect(c.GHInfo)
 
-	username := c.GHInfo.GitHubUsername
-	query := getYourAssigneeSearchQuery(username, config.GitHubOrg)
-	result, _, err := githubClient.Search.Issues(c.Ctx, query, &github.SearchOptions{})
+	reviewResp, assignmentResp, openPrResp, err := graphQLClient.GetLHSData()
```
Comment (ayusht2810) [inline server/plugin/api.go]:
```suggestion
	reviewResp, assignmentResp, openPRResp, err := graphQLClient.GetLHSData()
```

Comment (ayusht2810) [reply server/plugin/api.go]: maybe? change everywhere if changing?

File: server/plugin/api.go [hunk 3]
```diff
@@ -970,27 +938,24 @@ func (p *Plugin) createIssueComment(c *UserContext, w http.ResponseWriter, r *ht
 	p.writeJSON(w, result)
 }
 
-func (p *Plugin) getYourAssignmentsData(c *UserContext) []*github.Issue {
-	config := p.getConfiguration()
-
-	githubClient := p.githubConnectUser(c.Context.Ctx, c.GHInfo)
+func (p *Plugin) getLHSData(c *UserContext) ([]*github.Issue, []*github.Issue, []*github.Issue) {
```
Comment (ayusht2810) [inline server/plugin/api.go]: should we return here a list of `[]*github.Issue` instead of having the same parameters multiple times?

File: server/plugin/internal/graphql/client.go [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+	if enterpriseBaseURL == "" || orgName == "" {
+		client = Client{
+			username: username,
+			client:   githubv4.NewClient(httpClient),
+			api:      api,
+		}
+	} else {
+		baseURL, err := url.Parse(enterpriseBaseURL)
+		if err != nil {
+			api.LogDebug("Not able to parse the URL", "Error", err)
```
Comment (ayusht2810) [inline server/plugin/internal/graphql/client.go]:
```suggestion
			api.LogDebug("Not able to parse the URL", "Error", err.Error())
```

File: server/plugin/internal/graphql/lhs_request.go [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+	"github.com/shurcooL/githubv4"
+)
+
+const (
+	queryParamReviewCursor      = "reviewCursor"
+	queryParamAssignmentsCursor = "assignmentsCursor"
+	queryParamOpenPRsCursor     = "openPrsCursor"
+
+	queryParamOpenPRQueryArg   = "prOpenQueryArg"
+	queryParamReviewQueryArg   = "prReviewQueryArg"
```
Comment (ayusht2810) [inline server/plugin/internal/graphql/lhs_request.go]:
```suggestion
	queryParamReviewPRQueryArg   = "prReviewQueryArg"
```

File: server/plugin/internal/graphql/lhs_request.go [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+	}
+
+	if c.org != "" {
+		params[queryParamOpenPRQueryArg] = githubv4.String(fmt.Sprintf("org:%s %s", c.org, params[queryParamOpenPRQueryArg]))
+		params[queryParamReviewQueryArg] = githubv4.String(fmt.Sprintf("org:%s %s", c.org, params[queryParamReviewQueryArg]))
+		params[queryParamAssigneeQueryArg] = githubv4.String(fmt.Sprintf("org:%s %s", c.org, params[queryParamAssigneeQueryArg]))
+	}
+
+	var resultPR, resultAssignee, resultOpenPR []*github.Issue
+	flagPR, flagAssignee, flagOpenPr := false, false, false
```
Comment (ayusht2810) [inline server/plugin/internal/graphql/lhs_request.go]:
```suggestion
	flagPR, flagAssignee, flagOpenPR := false, false, false
```

File: server/plugin/internal/graphql/lhs_request.go [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+	queryParamReviewCursor      = "reviewCursor"
+	queryParamAssignmentsCursor = "assignmentsCursor"
+	queryParamOpenPRsCursor     = "openPrsCursor"
+
+	queryParamOpenPRQueryArg   = "prOpenQueryArg"
+	queryParamReviewQueryArg   = "prReviewQueryArg"
+	queryParamAssigneeQueryArg = "assigneeQueryArg"
+)
+
+func (c *Client) GetLHSData() ([]*github.Issue, []*github.Issue, []*github.Issue, error) {
```
Comment (ayusht2810) [inline server/plugin/internal/graphql/lhs_request.go:20]: same here for parameters as well

File: server/plugin/internal/graphql/lhs_request.go [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+			}
+
+			params[queryParamOpenPRsCursor] = githubv4.NewString(mainQuery.OpenPullRequest.PageInfo.EndCursor)
+		}
+	}
+
+	return resultPR, resultAssignee, resultOpenPR, nil
+}
+
+func getPRorIssue(prResp *prSearchNodes, assignmentResp *assignmentSearchNodes) *github.Issue {
```
Comment (ayusht2810) [inline server/plugin/internal/graphql/lhs_request.go]:
```suggestion
func getPROrIssue(prResp *prSearchNodes, assignmentResp *assignmentSearchNodes) *github.Issue {
```

File: go.mod (hunk 1 above)
Comment (Kshitij-Katiyar) [reply go.mod]: nope tried that

File: server/plugin/api.go (hunk 3 above)
Comment (Kshitij-Katiyar) [reply server/plugin/api.go]: discussed

**Review by ayusht2810 — APPROVED** [review]

File: server/plugin/api.go (hunk 3 above)
Comment (manojmalik20) [inline server/plugin/api.go]: We should also specify the names of the return variables here to understand what the function is returning.

File: server/plugin/internal/graphql/client.go [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+
+// Client encapsulates the third party package that communicates with Github GraphQL API
+type Client struct {
+	client   *githubv4.Client
+	org      string
+	username string
+	api      plugin.API
+}
+
+// NewClient creates and returns Client. Third party package that queries GraphQL is initialized here.
```
Comment (manojmalik20) [inline server/plugin/internal/graphql/client.go]:
```suggestion
// NewClient creates and returns Client. The third party package that queries GraphQL is initialized here.
```

File: server/plugin/internal/graphql/client.go [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+			client:   githubv4.NewClient(httpClient),
+			api:      api,
+		}
+	} else {
+		baseURL, err := url.Parse(enterpriseBaseURL)
+		if err != nil {
+			api.LogDebug("Not able to parse the URL", "Error", err.Error())
+		}
+
+		baseURL.Path = path.Join(baseURL.Path, "api", "graphql")
```
Comment (manojmalik20) [inline server/plugin/internal/graphql/client.go:41]: This line can have nil pointer dereference error as we are not returning in the above `if err != nil` condition. In that case, `baseURL` can be nil.

File: server/plugin/internal/graphql/lhs_request.go [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+			break
+		}
+
+		if err := c.executeQuery(&mainQuery, params); err != nil {
+			return nil, nil, nil, err
+		}
+
+		if !flagPR {
+			for _, resp := range mainQuery.PullRequest.Nodes {
+				resp := resp
```
Comment (manojmalik20) [inline server/plugin/internal/graphql/lhs_request.go:50]: What's the purpose of this line?

File: webapp/src/selectors.js [hunk 12]
```diff
@@ -52,11 +52,11 @@ function mapPrsToDetails(prs, details) {
 export const getSidebarData = createSelector(
     getPluginState,
     (pluginState) => {
-        const {username, sidebarContent, reviewDetails, yourPrDetails, organization, rhsState} = pluginState;
+        const {username, sidebarContent, reviewsDetails, yourPrsDetails, organization, rhsState} = pluginState;
```
Comment (manojmalik20) [inline webapp/src/selectors.js:55]: What's the reason for these changes?

File: server/plugin/internal/graphql/lhs_request.go (hunk 11 above)
Comment (Kshitij-Katiyar) [reply server/plugin/internal/graphql/lhs_request.go:50]: It was to fix the CI error ```G601: Implicit memory aliasing in for loop. (gosec)```

File: webapp/src/selectors.js (hunk 12 above)
Comment (Kshitij-Katiyar) [reply webapp/src/selectors.js:55]: Previously it was wrong now i have corrected it.

Comment (manojmalik20) [reply webapp/src/selectors.js:55]: I think the previous names were better.

**Review by manojmalik20 — APPROVED** [review]
